src/Avalonia.Base/Styling/Styles.cs

```diff
@@ -227,70 +230,75 @@ void IResourceProvider.RemoveOwner(IResourceHost owner)
             }
         }
 
-        private void OnCollectionChanged(object? sender, NotifyCollectionChangedEventArgs e)
+        private static IReadOnlyList<T> ToReadOnlyList<T>(ICollection list)
         {
-            static IReadOnlyList<T> ToReadOnlyList<T>(IList list)
+            if (list is IReadOnlyList<T> readOnlyList)
             {
-                if (list is IReadOnlyList<T>)
-                {
-                    return (IReadOnlyList<T>)list;
-                }
-                else
-                {
-                    var result = new T[list.Count];
-                    list.CopyTo(result, 0);
-                    return result;
-                }
+                return readOnlyList;
             }
 
-            void Add(IList items)
+            var result = new T[list.Count];
+            list.CopyTo(result, 0);
+            return result;
+        }
+
+        private static void InternalAdd(IList items, IResourceHost? owner, ref StyleCache? cache)
+        {
+            if (owner is not null)
             {
-                for (var i = 0; i < items.Count; ++i)
+                foreach (var resourceProvider in items.OfType<IResourceProvider>())
                 {
-                    var style = (IStyle)items[i]!;
-
-                    if (Owner is object && style is IResourceProvider resourceProvider)
-                    {
-                        resourceProvider.AddOwner(Owner);
-                    }
-
-                    _cache = null;
+                    resourceProvider.AddOwner(owner);
                 }
 
-                (Owner as IStyleHost)?.StylesAdded(ToReadOnlyList<IStyle>(items));
+                (owner as IStyleHost)?.StylesAdded(ToReadOnlyList<IStyle>(items));
+            }
+
+            if (items.Count > 0)
+            {
+                cache = null;
             }
+        }
 
-            void Remove(IList items)
+        private static void InternalRemove(IList items, IResourceHost? owner, ref StyleCache? cache)
+        {
+            if (owner is not null)
             {
-                for (var i = 0; i < items.Count; ++i)
+                foreach (var resourceProvider in items.OfType<IResourceProvider>())
                 {
-                    var style = (IStyle)items[i]!;
+                    resourceProvider.RemoveOwner(owner);
```

Comment:
We generally avoid using LINQ to minimize allocations.

```suggestion
                    if (items[i] is IResourceProvider resourceProvider)
                    {
                        resourceProvider.RemoveOwner(owner);
                    }
```

Reply:
Hello, thank you for taking the time to do the review.

Yes, you are basically right, if the method is called a million times then you should notice the difference.

That is again a code readability vs performance story.

If I search in the code of Avalonia for ".OfType<" I find 303 places :grin:.

Unfortunately I didn't find exact information what is standard here and what is not. At least I can't find anything about it at https://github.com/AvaloniaUI/Avalonia/blob/master/CONTRIBUTING.md.

Reply:
You wouldn't find Linq in perf critical places of the framework. Or if there are any - it should be fixed.